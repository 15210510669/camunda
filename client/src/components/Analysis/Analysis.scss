.Analysis {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
}

.Analysis__diagram {
  display: flex;
  flex-grow: 1;
}
