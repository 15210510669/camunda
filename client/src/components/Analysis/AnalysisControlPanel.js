/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import classnames from 'classnames';
import update from 'immutability-helper';
import React from 'react';
import equal from 'deep-equal';
import {ActionItem, DefinitionSelection} from 'components';

import {getFlowNodeNames} from 'services';

import {Filter} from '../Reports';

import './AnalysisControlPanel.scss';

export default class AnalysisControlPanel extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      flowNodeNames: null
    };
  }

  componentDidMount() {
    this.loadFlowNodeNames();
  }

  loadFlowNodeNames = async () => {
    this.setState({
      flowNodeNames: await getFlowNodeNames(
        this.props.processDefinitionKey,
        this.props.processDefinitionVersions[0],
        this.props.tenantIds[0]
      )
    });
  };

  componentDidUpdate({processDefinitionKey, processDefinitionVersions}) {
    if (
      this.props.processDefinitionKey !== processDefinitionKey ||
      !equal(this.props.processDefinitionVersions, processDefinitionVersions)
    ) {
      this.loadFlowNodeNames();
    }
  }

  getDefinitionConfig = () => {
    return {
      processDefinitionKey: this.props.processDefinitionKey,
      processDefinitionVersions: this.props.processDefinitionVersions,
      tenantIds: this.props.tenantIds
    };
  };

  hover = element => () => {
    this.props.updateHover(element);
  };

  isProcDefSelected = () => {
    return this.props.processDefinitionKey && this.props.processDefinitionVersions;
  };

  renderInput = ({type, label, bpmnKey}) => {
    const {hoveredControl, hoveredNode} = this.props;
    const disableFlowNodeSelection = !this.isProcDefSelected();

    return (
      <div
        className={classnames('AnalysisControlPanel__config', {
          'AnalysisControlPanel__config--hover':
            !disableFlowNodeSelection &&
            (hoveredControl === type || (hoveredNode && hoveredNode.$instanceOf(bpmnKey)))
        })}
        name={'AnalysisControlPanel__' + type}
        onMouseOver={this.hover(type)}
        onMouseOut={this.hover(null)}
      >
        <ActionItem
          disabled={disableFlowNodeSelection}
          onClick={() => this.props.updateSelection(type, null)}
        >
          {this.props[type] ? this.props[type].name || this.props[type].id : 'Select ' + label}
        </ActionItem>
      </div>
    );
  };

  render() {
    return (
      <div className="AnalysisControlPanel">
        <ul className="AnalysisControlPanel__list">
          <li className="AnalysisControlPanel__item summary">
            For
            <DefinitionSelection
              type="process"
              definitionKey={this.props.processDefinitionKey}
              versions={this.props.processDefinitionVersions}
              tenants={this.props.tenantIds}
              xml={this.props.xml}
              onChange={(key, versions, tenantIds) =>
                this.props.onChange({
                  processDefinitionKey: key,
                  processDefinitionVersions: versions,
                  tenantIds,
                  filter: this.props.filter.filter(
                    ({type}) => type !== 'executedFlowNodes' && type !== 'variable'
                  )
                })
              }
            />
            analyse how the branches of
            {this.renderInput({type: 'gateway', label: 'Gateway', bpmnKey: 'bpmn:Gateway'})}
            affect the probability that an instance reached
            {this.renderInput({type: 'endEvent', label: 'End Event', bpmnKey: 'bpmn:EndEvent'})}
          </li>
          <li className="AnalysisControlPanel__item AnalysisControlPanel__item--filter">
            <Filter
              data={this.props.filter}
              flowNodeNames={this.state.flowNodeNames}
              onChange={({filter}) =>
                this.props.onChange({filter: update(this.props.filter, filter)})
              }
              xml={this.props.xml}
              {...this.getDefinitionConfig()}
            />
          </li>
        </ul>
      </div>
    );
  }
}
