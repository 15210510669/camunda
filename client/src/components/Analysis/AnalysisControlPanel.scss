.AnalysisControlPanel {
  background-color: #f4f4f4;
  border-radius: 8px;
  border: 1px solid #dfdfdf;
  padding: 12px;
}

.AnalysisControlPanel__popover {
  display: inline-block;
  margin: 0 0.5em;
}

.AnalysisControlPanel__list {
  list-style: none;
  margin: 0;
  padding: 0;
  display: inline;
}

.AnalysisControlPanel__item {
  display: inline-block;
  position: relative;
  padding-right: 20px;

  .Labeled .label {
    color: #626262;
    font-size: 0.875em;
    margin-bottom: 5px;
  }
}

.AnalysisControlPanel__item--select {
  width: 15%;
  min-width: 200px;
}

.AnalysisControlPanel__item--filter {
  display: inline;
  padding-top: 27px;
}

.AnalysisControlPanel__select {
  width: 100%;
}

.AnalysisControlPanel__config {
  cursor: default;
  display: inline-block;
  margin: 0 0.5em;
}

.AnalysisControlPanel__config--hover .ActionItem {
  background-color: rgba(48, 116, 142, 0.6);
}
