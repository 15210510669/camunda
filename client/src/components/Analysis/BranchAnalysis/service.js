/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import {post} from 'request';

export async function loadFrequencyData(
  processDefinitionKey,
  processDefinitionVersions,
  tenantIds,
  filter
) {
  const response = await post(
    'api/report/evaluate',
    createFlowNodeFrequencyReport(
      processDefinitionKey,
      processDefinitionVersions,
      tenantIds,
      filter
    )
  );

  return await response.json();
}

function createFlowNodeFrequencyReport(
  processDefinitionKey,
  processDefinitionVersions,
  tenantIds,
  filter
) {
  return {
    combined: false,
    reportType: 'process',
    data: {
      processDefinitionKey,
      processDefinitionVersions,
      tenantIds,
      filter,
      view: {
        entity: 'flowNode',
        property: 'frequency',
      },
      groupBy: {
        type: 'flowNodes',
        unit: null,
      },
      visualization: 'heat',
    },
  };
}

export async function loadCorrelationData(
  processDefinitionKey,
  processDefinitionVersions,
  tenantIds,
  filter,
  gateway,
  end
) {
  const response = await post('api/analysis/correlation', {
    processDefinitionKey,
    processDefinitionVersions,
    tenantIds,
    filter,
    gateway,
    end,
  });

  return await response.json();
}
