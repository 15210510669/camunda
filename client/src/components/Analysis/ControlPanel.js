import React from 'react';
import {ActionItem, Popover, ProcessDefinitionSelection} from 'components';

import {extractProcessDefinitionName} from 'services';

import {Filter} from '../Reports';

import './ControlPanel.css';

export default class ControlPanel extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      processDefinitionName: this.props.processDefinitionKey
    }
  }

  loadProcessDefinitionName = async (xml) => {
    const {processDefinitionName} = await extractProcessDefinitionName(xml);
    this.setState({
      processDefinitionName
    })
  }

  componentWillReceiveProps(nextProps) {
    if(this.props.xml !== nextProps.xml) {
      this.loadProcessDefinitionName(nextProps.xml);
    }
  }

  getDefinitionConfig = () => {
    return {
      processDefinitionKey: this.props.processDefinitionKey,
      processDefinitionVersion: this.props.processDefinitionVersion
    }
  }

  hover = element => () => {
    this.props.updateHover(element);
  }

  createTitle = () => {
    const {processDefinitionKey, processDefinitionVersion} = this.props;
    const processDefintionName = this.state.processDefinitionName? this.state.processDefinitionName : processDefinitionKey;
    if(processDefintionName && processDefinitionVersion) {
      return `${processDefintionName} : ${processDefinitionVersion}`;
    } else {
      return 'Select Process Definition';
    }
  }

  render() {
    const {hoveredControl, hoveredNode} = this.props;

    return (<div className='ControlPanel'>
        <ul className='ControlPanel__list'>
          <li className='ControlPanel__item ControlPanel__item--select'>
            <label htmlFor='ControlPanel__process-definition' className='ControlPanel__label'>Process definition</label>
            <Popover className='ControlPanel__popover' title={this.createTitle()}>
              <ProcessDefinitionSelection {...this.getDefinitionConfig()} xml={this.props.xml} onChange={this.props.onChange} />
            </Popover>
          </li>
          {[
          {type: 'endEvent', label: 'End Event', bpmnKey:'bpmn:EndEvent'},
          {type: 'gateway', label: 'Gateway', bpmnKey:'bpmn:Gateway'}].map(({type, label, bpmnKey}) =>
            <li key={type} className='ControlPanel__item'>
              <label htmlFor={'ControlPanel__' + type} className='ControlPanel__label'>{label}</label>
              <div
                className={'ControlPanel__config' + (hoveredControl === type || (hoveredNode && hoveredNode.$instanceOf(bpmnKey)) ? ' ControlPanel__config--hover' : '')}
                name={'ControlPanel__' + type}
                onMouseOver={this.hover(type)}
                onMouseOut={this.hover(null)}
                >
                  <ActionItem onClick={() => this.props.updateSelection(type, null)}>
                    {this.props[type] ?
                      this.props[type].name || this.props[type].id
                    : 'Please Select ' + label}
                  </ActionItem>
              </div>
            </li>
          )}
          <li className='ControlPanel__item ControlPanel__item--filter'>
            <Filter data={this.props.filter} onChange={this.props.onChange} {...this.getDefinitionConfig()} />
          </li>
        </ul>
      </div>
    );
  }

}

