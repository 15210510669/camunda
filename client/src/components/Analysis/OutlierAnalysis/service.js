/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {get} from 'request';

export async function loadNodesOutliers(config) {
  const response = await get('api/analysis/flowNodeOutliers', config);
  return await response.json();
}

export {loadDurationData, loadCommonOutliersVariables, getInstancesDownloadUrl} from './service.ts';
