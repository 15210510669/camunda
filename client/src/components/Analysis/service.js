import {get, post} from 'request';

export async function loadProcessDefinitionXml(processDefinitionKey, processDefinitionVersion) {
  const response = await get('/api/process-definition/xml', {processDefinitionKey, processDefinitionVersion});

  return await response.text();
}

export async function loadFrequencyData(processDefinitionKey, processDefinitionVersion, filter) {
  const response = await post('/api/process-definition/heatmap/frequency', {
    processDefinitionKey,
    processDefinitionVersion,
    filter
  });

  return await response.json();
}

export async function loadCorrelationData(processDefinitionKey, processDefinitionVersion, filter, gateway, end) {
  const response = await post('/api/process-definition/correlation', {
    processDefinitionKey,
    processDefinitionVersion,
    filter,
    gateway,
    end
  });

  return await response.json();
}
