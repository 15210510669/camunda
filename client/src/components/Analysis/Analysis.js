import React from 'react';

import ControlPanel from './ControlPanel';
import {BPMNDiagram} from 'components';

import {loadProcessDefinitionXml, loadFrequencyData} from './service';
import DiagramBehavior from './DiagramBehavior';
import Statistics from './Statistics';

import './Analysis.css';

export default class Analysis extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      config: {
        processDefinitionId: '',
        processDefinitionKey: '',
        processDefinitionVersion: '',
        filter: []
      },
      data: null,
      hoveredControl: null,
      hoveredNode: null,
      gateway: null,
      endEvent: null,
      xml: null
    };
  }

  render() {
    const {
      xml,
      config,
      hoveredControl,
      hoveredNode,
      gateway,
      endEvent,
      data
    } = this.state;

    return <div className='Analysis'>
      <ControlPanel
        {...config}
        hoveredControl={hoveredControl}
        hoveredNode={hoveredNode}
        onChange={this.updateConfig}
        gateway={gateway}
        endEvent={endEvent}
        updateHover={this.updateHoveredControl}
        updateSelection={this.updateSelection}
        xml={xml}
        />
      <div className='Analysis__diagram'>
        {xml && <BPMNDiagram xml={xml}>
          <DiagramBehavior
            hoveredControl={hoveredControl}
            hoveredNode={hoveredNode}
            updateHover={this.updateHoveredNode}
            updateSelection={this.updateSelection}
            gateway={gateway}
            endEvent={endEvent}
            data={data}
          />
        </BPMNDiagram>}
      </div>
      {
        gateway && endEvent &&
        <Statistics
          gateway={gateway}
          endEvent={endEvent}
          config={config}
        />
      }
    </div>;
  }

  async componentDidUpdate(prevProps, prevState) {
    const {config} = this.state;
    const {config: prevConfig} = prevState;
    if(config.processDefinitionId &&
        (prevConfig.filter !== config.filter ||
         prevConfig.processDefinitionId !== config.processDefinitionId)) {

      this.setState({data: await loadFrequencyData(config.processDefinitionId, config.filter)});
    }
  }

  updateHoveredControl = newField => {
    this.setState({hoveredControl: newField});
  }

  updateHoveredNode = newNode => {
    this.setState({hoveredNode: newNode});
  }

  updateSelection = (type, node) => {
    this.setState({[type]: node});
  }

  updateConfig = async (updates) => {
    const config = {
      ...this.state.config
    };
    updates.forEach( ({field, newValue}) => {
      config[field] = newValue;
    });
    this.setState({config});

    const processDefinitionIdWasUpdated = updates.find(({field}) => field === 'processDefinitionId');
    if(processDefinitionIdWasUpdated && processDefinitionIdWasUpdated.newValue) {
      this.setState({
        xml: await loadProcessDefinitionXml(processDefinitionIdWasUpdated.newValue),
        gateway: null,
        endEvent: null
      });
    }
  }
}
