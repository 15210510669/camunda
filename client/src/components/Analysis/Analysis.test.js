/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';
import {mount, shallow} from 'enzyme';

import {loadFrequencyData} from './service';

import Analysis from './Analysis';

import {incompatibleFilters, loadProcessDefinitionXml} from 'services';

jest.mock('./AnalysisControlPanel', () => () => <div>ControlPanel</div>);

jest.mock('components', () => {
  return {
    BPMNDiagram: () => <div>BPMNDiagram</div>,
    Message: () => <div>Message</div>
  };
});

jest.mock('./service', () => {
  return {
    loadFrequencyData: jest.fn()
  };
});

jest.mock('services', () => {
  return {
    loadProcessDefinitionXml: jest.fn(),
    loadDefinitions: () => [{key: 'key', versions: [{version: 2}, {version: 1}]}],
    incompatibleFilters: jest.fn()
  };
});

jest.mock('./DiagramBehavior', () => () => <div>DiagramBehavior</div>);
jest.mock('./Statistics', () => () => <div>Statistics</div>);

it('should contain a control panel', () => {
  const node = mount(<Analysis />);

  expect(node).toIncludeText('ControlPanel');
});

it('should load the process definition xml when the process definition id is updated', () => {
  const node = mount(<Analysis />);

  loadProcessDefinitionXml.mockClear();
  node.instance().updateConfig({
    processDefinitionKey: 'someKey',
    processDefinitionVersions: ['someVersion'],
    tenantIds: ['a', 'b']
  });

  expect(loadProcessDefinitionXml).toHaveBeenCalledWith('someKey', 'someVersion', 'a');
});

it('should load frequency data when the process definition key changes', async () => {
  const node = mount(<Analysis />);

  node
    .instance()
    .updateConfig({processDefinitionKey: 'someKey', processDefinitionVersions: ['someVersion']});
  loadFrequencyData.mockClear();
  await node.instance().updateConfig({processDefinitionKey: 'anotherKey'});

  expect(loadFrequencyData.mock.calls[0][0]).toBe('anotherKey');
});

it('should load frequency data when the process definition version changes', async () => {
  const node = mount(<Analysis />);

  await node
    .instance()
    .updateConfig({processDefinitionKey: 'someKey', processDefinitionVersions: ['someVersion']});
  loadFrequencyData.mockClear();
  await node.instance().updateConfig({processDefinitionVersions: ['anotherVersion']});

  expect(loadFrequencyData.mock.calls[0][1]).toEqual(['anotherVersion']);
});

it('should load updated frequency data when the filter changed', async () => {
  const node = mount(<Analysis />);

  await node.instance().updateConfig({
    processDefinitionKey: 'someKey',
    processDefinitionVersions: ['someVersion'],
    tenantIds: [null]
  });
  loadFrequencyData.mockClear();
  await node.instance().updateConfig({filter: ['someFilter']});

  expect(loadFrequencyData.mock.calls[0][3]).toEqual(['someFilter']);
});

it('should not try to load frequency data if no process definition is selected', () => {
  const node = mount(<Analysis />);

  loadFrequencyData.mockClear();
  node.instance().updateConfig({filter: ['someFilter']});

  expect(loadFrequencyData).not.toHaveBeenCalled();
});

it('should contain a statistics section if gateway and endEvent is selected', () => {
  const node = mount(<Analysis />);

  node.instance().setState({
    gateway: 'g',
    endEvent: 'e'
  });

  expect(node).toIncludeText('Statistics');
});

it.only('should clear the selection when another process definition is selected', async () => {
  const node = mount(<Analysis />);

  await node.instance().setState({gateway: 'g', endEvent: 'e'});
  await node.instance().updateConfig({
    processDefinitionKey: 'newKey',
    processDefinitionVersions: ['latest'],
    tenantIds: []
  });

  expect(node).toHaveState('gateway', null);
  expect(node).toHaveState('endEvent', null);
  expect(node.state().config.filter).toEqual([]);
});

it('should show a warning message when there are incompatible filters', async () => {
  incompatibleFilters.mockReturnValue(true);
  const node = await mount(<Analysis />);
  await node.update();
  expect(node.find('Message')).toExist();
});

it('should not reset the xml when adding a filter', async () => {
  const node = shallow(<Analysis />);

  await node.instance().updateConfig({
    processDefinitionKey: 'someKey',
    processDefinitionVersions: ['someVersion'],
    tenantIds: ['a', 'b']
  });
  await node.instance().updateConfig({
    filter: [{type: 'completedInstancesOnly', data: null}]
  });

  expect(node.state().xml).not.toBe(null);
});
