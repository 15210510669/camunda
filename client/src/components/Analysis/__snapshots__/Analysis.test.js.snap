// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render sub navigation 1`] = `
<div
  className="Analysis"
>
  <SubNav>
    <SubNavItem
      active={
        Array [
          "/analysis",
          "/analysis/taskAnalysis",
        ]
      }
      linksTo="/analysis/taskAnalysis"
      name="Task Analysis"
    />
    <SubNavItem
      active={
        Array [
          "/analysis/branchAnalysis",
        ]
      }
      linksTo="/analysis/branchAnalysis"
      name="Branch analysis"
    />
  </SubNav>
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/analysis/"
    />
    <Route
      component={[Function]}
      path="/analysis/branchAnalysis"
    />
    <Route
      component={[Function]}
      path="/analysis/taskAnalysis"
    />
    <Route
      component={[Function]}
      path="*"
    />
  </Switch>
</div>
`;
